gql/comparator.go

```diff
@@ -0,0 +1,207 @@
+package gql
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+)
+
+// A Comparator checks whether a value Matches.
+// Each NewListOperation returns a comparator to be used in List filtering.
+type Comparator interface {
+	Match(interface{}) bool
+}
+
+// NotComparator simply does a logical not on the Match result of its child.
+type NotComparator struct {
+	Child Comparator
+}
+
+func (c NotComparator) Match(raw interface{}) bool {
+	return !c.Child.Match(raw)
+}
+
+// NewEqualComparator returns a comparator for equality of both strings and ints.
+// The comparator expects either a string or int with the key "Value" as part of the given argument.
+func NewEqualComparator(arg map[string]interface{}) (Comparator, error) {
+	raw, ok := arg["Value"]
+	if !ok {
+		return nil, errors.New("filter argument is missing 'Value'")
+	}
+	switch a := raw.(type) {
+	case int:
+		return intEqual{value: a}, nil
+	case string:
+		return stringEqual{value: a}, nil
+	default:
+		return nil, errors.New("unsupported argument value, strings and integers are supported")
+	}
+}
+
+// NewNotEqualComparator returns a comparator for not equality of both strings and ints.
+// The comparator expects either a string or int with the key "Value" as part of the given argument.
+func NewNotEqualComparator(arg map[string]interface{}) (Comparator, error) {
+	eq, err := NewEqualComparator(arg)
+	if err != nil {
+		return nil, err
+	}
+	return NotComparator{Child: eq}, nil
+}
+
+// NewInComparator returns a comparator which returns true if the operand matches any of the given values.
+// The comparator expects a list of strings under the key "Values" as part of the given argument.
+func NewInComparator(arg map[string]interface{}) (Comparator, error) {
+	raw, ok := arg["Values"]
+	if !ok {
+		return nil, errors.New("filter argument is missing 'Values'")
+	}
+	value := reflect.ValueOf(raw)
+	if value.Kind() != reflect.Slice {
+		return nil, errors.New("filter argument 'Values' should be a list of strings")
+	}
+
+	values := make([]string, value.Len())
+	for i := 0; i < value.Len(); i++ {
+		rawItem := value.Index(i).Interface()
+		item, ok := rawItem.(string)
+		if !ok {
+			return nil, errors.New("filter argument 'Values' should be a list of strings")
+		}
+		values[i] = item
+	}
+
+	return inComparator{values: values}, nil
+}
+
+// NewIntegerComparator returns a NewListOperator for integer operations supported by integerComparator,
+// specifically <, <=, > and >=.
+// The returned NewListOperation expects an int with the key "Value" as part of the given argument.
+func NewIntegerComparator(op string) NewListOperation {
```

Comment:
Maybe I'll find the reason as I continue reading, but this one is an odd duck compared to the others in that it returns `NewListOperation` instead of `Comparator`, which makes its error handling really weird.

Reply:
Yeah it is different but saves on a lot of repeated code.

Reply:
good point, fixing